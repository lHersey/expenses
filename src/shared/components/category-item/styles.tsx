import Text from 'shared/components/text';
import { colorAdjust } from 'shared/hooks';
import styled from 'styled-components/native';

export const Container = styled.Pressable`
  background: ${p => p.theme.SECONDARY_BACKGROUND_COLOR};
  border-radius: 20px;
  padding: 10px;
  align-items: center;
  flex-grow: 1;
  flex-direction: row;
  margin-top: 10px;
`;

export const IconContainer = styled.View<{ color: string }>`
  width: 50px;
  height: 50px;
  margin-right: 20px;
  border-radius: 25px;
  background: ${p => colorAdjust(p.color, 30)};
  justify-content: center;
  align-items: center;
`;

export const Label = styled(Text)`
  font-family: Inter_300;
`;
