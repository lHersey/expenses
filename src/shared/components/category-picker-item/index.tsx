import React, { FC } from 'react';
import { useModalState } from 'shared/hooks';
import { useSelector } from 'shared/store/hooks';
import { categoriesSelector } from 'shared/store/reducers/category/reducer';

import CategoryPickerModal from '../category-picker-modal';
import CreateTransactionItem from '../create-transaction-item';

type Props = {
  value?: string | null;
  onAccept: (date: string) => void;
};

const CategoryPickerItem: FC<Props> = props => {
  const { value, onAccept } = props;
  const [isVisible, handleToggle] = useModalState();
  const category = useSelector(state => categoriesSelector.selectById(state, value ?? ''));

  const handleOnAccept = (categoryId: string) => {
    handleToggle();
    onAccept(categoryId);
  };

  return (
    <>
      <CategoryPickerModal isVisible={isVisible} value={value} onAccept={handleOnAccept} />
      <CreateTransactionItem onPress={handleToggle} label={category?.name ?? 'Select category'} icon="apps" />
    </>
  );
};

export default CategoryPickerItem;
