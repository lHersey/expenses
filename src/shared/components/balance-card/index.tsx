import React, { FC } from 'react';
import { BalanceType } from 'shared/enum';

import BalanceTypeCard from '../balance-type-card';
import { AmmountLabel, BalanceTypeContainer, CardContainer, TextContainer, Title } from './styles';

const BalanceCard: FC = () => {
  return (
    <CardContainer gradient>
      <TextContainer>
        <Title>Total Balance</Title>
      </TextContainer>
      <TextContainer>
        <AmmountLabel>$ 4800.00</AmmountLabel>
      </TextContainer>
      <BalanceTypeContainer>
        <BalanceTypeCard type={BalanceType.INCOME} ammount={4309} />
        <BalanceTypeCard type={BalanceType.EXPENSE} ammount={400} />
      </BalanceTypeContainer>
    </CardContainer>
  );
};

export default BalanceCard;
