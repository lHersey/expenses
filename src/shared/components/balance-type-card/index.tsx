import { Ionicons } from '@expo/vector-icons';
import { FC } from 'react';
import { PressableProps } from 'react-native';
import { BalanceType } from 'shared/enum';

import { useBalanceTypeInformation } from './hooks';
import { BalanceAmmount, BalanceLabelsContainer, BalanceTitle, Pressable } from './styles';

type Props = PressableProps & {
  type: BalanceType;
  ammount: number;
};

const BalanceTypeCard: FC<Props> = props => {
  const { ammount, type } = props;
  const balanceInfo = useBalanceTypeInformation(type);

  return (
    <Pressable>
      <Ionicons name={balanceInfo?.icon} color={balanceInfo?.color} size={32} />
      <BalanceLabelsContainer>
        <BalanceTitle>{balanceInfo?.label}</BalanceTitle>
        <BalanceAmmount>${ammount}.00</BalanceAmmount>
      </BalanceLabelsContainer>
    </Pressable>
  );
};

export default BalanceTypeCard;
