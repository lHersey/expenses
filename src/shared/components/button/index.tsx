import React, { FC } from 'react';
import { IoniconsName } from 'shared/types/icons';
import { useTheme } from 'styled-components/native';

import { Icon, Label, LabelsContainer, Modifiers, Pressable, Subtitle } from './styles';

type Props = Modifiers & {
  text: string;
  subtitle?: string;
  icon?: IoniconsName;
};

const Button: FC<Props> = props => {
  const { text, icon, subtitle, ...modifiers } = props;
  const theme = useTheme();

  return (
    <Pressable {...modifiers}>
      <LabelsContainer>
        {subtitle && <Subtitle>{subtitle}</Subtitle>}
        <Label {...modifiers}>{text}</Label>
      </LabelsContainer>
      {icon && <Icon name={icon} color={theme.WHITE} size={25} />}
    </Pressable>
  );
};

export default Button;
