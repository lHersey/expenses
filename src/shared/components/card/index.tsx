import { FC } from 'react';
import { ViewProps } from 'react-native';
import { useTheme } from 'styled-components/native';

import { Container, Gradient, ShadowContainer } from './styles';

type Props = ViewProps & {
  gradient?: boolean;
};

//TODO: Remove as never when the styles types are fixed

const Card: FC<Props> = props => {
  const { gradient, children, ...viewProps } = props;
  const { GRADIENT_BACKGROUND } = useTheme();

  if (gradient)
    return (
      <ShadowContainer {...(viewProps as never)}>
        <Gradient {...(viewProps as never)} colors={GRADIENT_BACKGROUND} start={{ x: 0, y: 0 }} end={{ x: 1, y: 0 }}>
          {children}
        </Gradient>
      </ShadowContainer>
    );

  return (
    <ShadowContainer {...(viewProps as never)}>
      <Container {...(viewProps as never)}>{children}</Container>
    </ShadowContainer>
  );
};

export default Card;
