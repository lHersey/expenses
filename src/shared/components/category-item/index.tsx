import { Ionicons } from '@expo/vector-icons';
import { FC } from 'react';
import { useCallback } from 'react';
import { colorAdjust } from 'shared/hooks';
import { useSelector } from 'shared/store/hooks';
import { categoriesSelector } from 'shared/store/reducers/category/reducer';

import { Container, IconContainer, Label } from './styles';

type Props = {
  categoryId: string;
  onPress: (id?: string) => void;
};

const CategoryItem: FC<Props> = props => {
  const { categoryId, onPress } = props;
  const category = useSelector(state => categoriesSelector.selectById(state, categoryId));

  const handleOnPress = useCallback(() => onPress(categoryId), [categoryId, onPress]);

  if (!category) return null;

  return (
    <Container onPress={handleOnPress} testID={CATEGORY_ITEM_TEST_ID}>
      <IconContainer color={category.color}>
        <Ionicons name={category.icon} color={colorAdjust(category.color, -30)} size={30} />
      </IconContainer>
      <Label>{category.name}</Label>
    </Container>
  );
};

export const CATEGORY_ITEM_TEST_ID = 'CATEGORY_ITEM';

export default CategoryItem;
