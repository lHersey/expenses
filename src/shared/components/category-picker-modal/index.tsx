import { FC } from 'react';
import { ScrollView } from 'react-native-gesture-handler';
import Modal from 'react-native-modal';
import { useSelector } from 'shared/store/hooks';
import { categoriesSelector } from 'shared/store/reducers/category/reducer';

import CategoryItem from '../category-item';
import { Container } from './styles';

type Props = {
  isVisible: boolean;
  value?: string | null;
  onAccept: (value: string) => void;
};

const CategoryPickerModal: FC<Props> = props => {
  const { isVisible, onAccept } = props;
  const categories = useSelector(state => categoriesSelector.selectIds(state));

  return (
    <Modal isVisible={isVisible} testID={CATEGORY_PICKER_MODAL_TEST_ID}>
      <Container>
        <ScrollView>
          {categories.map(id => (
            <CategoryItem key={id} categoryId={id.toString()} onPress={onAccept as never} />
          ))}
        </ScrollView>
      </Container>
    </Modal>
  );
};

export const CATEGORY_PICKER_MODAL_TEST_ID = 'CATEGORY_PICKER_MODAL_TEST_ID';

export default CategoryPickerModal;
