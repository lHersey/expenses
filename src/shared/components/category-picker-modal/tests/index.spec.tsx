import { fireEvent, render } from '@testing-library/react-native';
import { CATEGORY_ITEM_TEST_ID } from 'shared/components/category-item';
import Providers from 'shared/providers/test-providers';
import { mockCategory, mockCategoryEntity } from 'shared/store/mock';

import CategoryPickerModal from '../index';

describe('<CategoryPickerModal />', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  describe('snapshots', () => {
    it('should match snapshot on visible modal', () => {
      expect.hasAssertions();

      const categoryOne = mockCategory({ name: 'CategoryOne', id: '1' });
      const categoryTwo = mockCategory({ name: 'CategoryTwo', id: '2' });
      const categoryThree = mockCategory({ name: 'CategoryThree', id: '3' });

      const state = {
        category: mockCategoryEntity(categoryOne, categoryTwo, categoryThree),
      };

      const rendered = render(
        <Providers state={state}>
          <CategoryPickerModal isVisible value={null} onAccept={jest.fn()} />
        </Providers>,
      ).toJSON();

      expect(rendered).toMatchSnapshot();
    });

    it('should match snapshot with no visible modal', () => {
      expect.hasAssertions();

      const state = {
        category: mockCategoryEntity(),
      };

      const rendered = render(
        <Providers state={state}>
          <CategoryPickerModal isVisible={false} value={null} onAccept={jest.fn()} />
        </Providers>,
      ).toJSON();

      expect(rendered).toMatchSnapshot();
    });
  });

  describe('unit', () => {
    it('should render the text of each category', () => {
      expect.hasAssertions();

      const categoryOne = mockCategory({ name: 'CategoryOne', id: '1' });
      const categoryTwo = mockCategory({ name: 'CategoryTwo', id: '2' });
      const categoryThree = mockCategory({ name: 'CategoryThree', id: '3' });

      const state = {
        category: mockCategoryEntity(categoryOne, categoryTwo, categoryThree),
      };

      const { queryByText } = render(
        <Providers state={state}>
          <CategoryPickerModal isVisible value={null} onAccept={jest.fn()} />
        </Providers>,
      );

      expect(queryByText(categoryOne.name)).toBeTruthy();
      expect(queryByText(categoryTwo.name)).toBeTruthy();
      expect(queryByText(categoryThree.name)).toBeTruthy();
    });

    it('should render all the categories', () => {
      expect.hasAssertions();

      const categoryOne = mockCategory({ name: 'CategoryOne', id: '1' });
      const categoryTwo = mockCategory({ name: 'CategoryTwo', id: '2' });
      const categoryThree = mockCategory({ name: 'CategoryThree', id: '3' });

      const state = {
        category: mockCategoryEntity(categoryOne, categoryTwo, categoryThree),
      };

      const { queryAllByTestId } = render(
        <Providers state={state}>
          <CategoryPickerModal isVisible value={null} onAccept={jest.fn()} />
        </Providers>,
      );

      expect(queryAllByTestId(CATEGORY_ITEM_TEST_ID)).toHaveLength(3);
    });

    it('should handle onPress', () => {
      expect.hasAssertions();

      const categoryOne = mockCategory({ name: 'CategoryOne', id: '1' });
      const categoryTwo = mockCategory({ name: 'CategoryTwo', id: '2' });
      const categoryThree = mockCategory({ name: 'CategoryThree', id: '3' });

      const state = {
        category: mockCategoryEntity(categoryOne, categoryTwo, categoryThree),
      };

      const handleOnAcceptMock = jest.fn();

      const { getByText } = render(
        <Providers state={state}>
          <CategoryPickerModal isVisible value={null} onAccept={handleOnAcceptMock} />
        </Providers>,
      );

      const pressable = getByText(categoryTwo.name);
      fireEvent.press(pressable);

      expect(handleOnAcceptMock).toHaveBeenCalledWith(categoryTwo.id);
    });
  });
});
