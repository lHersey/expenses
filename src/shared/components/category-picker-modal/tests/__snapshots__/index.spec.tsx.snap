// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryPickerModal /> snapshots should match snapshot on visible modal 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
    testID="CATEGORY_PICKER_MODAL_TEST_ID"
    transparent={true}
    visible={true}
  >
    <View
      accessible={true}
      focusable={true}
      forwardedRef={[Function]}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "height": 1334,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 750,
        }
      }
      useNativeDriver={false}
    />
    <View
      deviceHeight={null}
      deviceWidth={null}
      forwardedRef={[Function]}
      hideModalContentWhileAnimating={false}
      onBackdropPress={[Function]}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      panResponderThreshold={4}
      pointerEvents="box-none"
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      statusBarTranslucent={false}
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
          "margin": 37.5,
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        }
      }
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeThreshold={100}
      useNativeDriver={false}
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#FFF",
              "borderRadius": 20,
              "display": "flex",
              "flexDirection": "column",
              "justifyContent": "center",
              "maxHeight": "80%",
              "paddingBottom": 20,
              "paddingLeft": 20,
              "paddingRight": 20,
              "paddingTop": 20,
            },
          ]
        }
      >
        <RCTScrollView
          collapsable={false}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
        >
          <View>
            <View
              accessible={true}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#f2f5f7",
                    "borderRadius": 20,
                    "flexDirection": "row",
                    "flexGrow": 1,
                    "marginTop": 10,
                    "paddingBottom": 10,
                    "paddingLeft": 10,
                    "paddingRight": 10,
                    "paddingTop": 10,
                  },
                ]
              }
              testID="CATEGORY_ITEM"
            >
              <View
                color="#d3fae2"
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#f1ffff",
                      "borderRadius": 25,
                      "height": 50,
                      "justifyContent": "center",
                      "marginRight": 20,
                      "width": 50,
                    },
                  ]
                }
              >
                <Text />
              </View>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#5e5e5e",
                      "fontFamily": "Inter_500",
                      "fontSize": 16,
                    },
                    Object {
                      "fontFamily": "Inter_300",
                    },
                  ]
                }
              >
                CategoryOne
              </Text>
            </View>
            <View
              accessible={true}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#f2f5f7",
                    "borderRadius": 20,
                    "flexDirection": "row",
                    "flexGrow": 1,
                    "marginTop": 10,
                    "paddingBottom": 10,
                    "paddingLeft": 10,
                    "paddingRight": 10,
                    "paddingTop": 10,
                  },
                ]
              }
              testID="CATEGORY_ITEM"
            >
              <View
                color="#d3fae2"
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#f1ffff",
                      "borderRadius": 25,
                      "height": 50,
                      "justifyContent": "center",
                      "marginRight": 20,
                      "width": 50,
                    },
                  ]
                }
              >
                <Text />
              </View>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#5e5e5e",
                      "fontFamily": "Inter_500",
                      "fontSize": 16,
                    },
                    Object {
                      "fontFamily": "Inter_300",
                    },
                  ]
                }
              >
                CategoryTwo
              </Text>
            </View>
            <View
              accessible={true}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#f2f5f7",
                    "borderRadius": 20,
                    "flexDirection": "row",
                    "flexGrow": 1,
                    "marginTop": 10,
                    "paddingBottom": 10,
                    "paddingLeft": 10,
                    "paddingRight": 10,
                    "paddingTop": 10,
                  },
                ]
              }
              testID="CATEGORY_ITEM"
            >
              <View
                color="#d3fae2"
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#f1ffff",
                      "borderRadius": 25,
                      "height": 50,
                      "justifyContent": "center",
                      "marginRight": 20,
                      "width": 50,
                    },
                  ]
                }
              >
                <Text />
              </View>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#5e5e5e",
                      "fontFamily": "Inter_500",
                      "fontSize": 16,
                    },
                    Object {
                      "fontFamily": "Inter_300",
                    },
                  ]
                }
              >
                CategoryThree
              </Text>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </Modal>
</RNCSafeAreaProvider>
`;

exports[`<CategoryPickerModal /> snapshots should match snapshot with no visible modal 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
    testID="CATEGORY_PICKER_MODAL_TEST_ID"
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      focusable={true}
      forwardedRef={[Function]}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "height": 1334,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 750,
        }
      }
      useNativeDriver={false}
    />
    <View
      deviceHeight={null}
      deviceWidth={null}
      forwardedRef={[Function]}
      hideModalContentWhileAnimating={false}
      onBackdropPress={[Function]}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      panResponderThreshold={4}
      pointerEvents="box-none"
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      statusBarTranslucent={false}
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
          "margin": 37.5,
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        }
      }
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeThreshold={100}
      useNativeDriver={false}
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#FFF",
              "borderRadius": 20,
              "display": "flex",
              "flexDirection": "column",
              "justifyContent": "center",
              "maxHeight": "80%",
              "paddingBottom": 20,
              "paddingLeft": 20,
              "paddingRight": 20,
              "paddingTop": 20,
            },
          ]
        }
      >
        <RCTScrollView
          collapsable={false}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
        >
          <View />
        </RCTScrollView>
      </View>
    </View>
  </Modal>
</RNCSafeAreaProvider>
`;
